Sources/ParseSwift/Authentication/3rd Party/ParseFacebook.swift

```diff
@@ -0,0 +1,251 @@
+//
+//  ParseFacebook.swift
+//  ParseSwift
+//
+//  Created by Abdulaziz Alhomaidhi on 3/18/21.
+//  Copyright © 2021 Parse Community. All rights reserved.
+
+import Foundation
+#if canImport(Combine)
+import Combine
+#endif
+
+// swiftlint:disable line_length
+
+/**
+ Provides utility functions for working with Facebook User Authentication and `ParseUser`'s.
+ Be sure your Parse Server is configured for [sign in with Facebook](https://docs.parseplatform.org/parse-server/guide/#configuring-parse-server-for-sign-in-with-facebook).
+ For information on acquiring Facebook sign-in credentials to use with `ParseFacebook`, refer to [Facebook's Documentation](https://developers.facebook.com/docs/facebook-login/limited-login.
+ */
+public struct ParseFacebook<AuthenticatedUser: ParseUser>: ParseAuthentication {
+
+    /// Authentication keys required for Facebook authentication.
+    enum AuthenticationKeys: String, Codable {
+        case id // swiftlint:disable:this identifier_name
+        case token
```

Comment:
You can leave `token`, but this is for limited login. You need to also add the original login `case accessToken`.

You can add coding keys right after the case statements since SwiftLint won't like `access_token`:

```
enum CodingKeys: String, CodingKey {
            case id // swiftlint:disable:this identifier_name 
            case accessToken = "access_token"
}
```

Reply:
From the docs, https://docs.parseplatform.org/parse-server/guide/#facebook-authdata it looks like Facebook also takes an `expirationData` which seems it should be a required field (I need to verify). You should add this and the coding keys will look like:

```swift
enum CodingKeys: String, CodingKey { 
  case id // swiftlint:disable:this identifier_name 
  case accessToken = "access_token" 
  case expirationData = "expiration_date"
}
```

Reply:
You will need to convert the `expirationDate` to the correct version of string. After the `enum AuthenticationKeys` declaration in ParseFacebook add the following computed property:

```swift
static let dateFormatter: DateFormatter = {
    var dateFormatter = DateFormatter()
    dateFormatter.locale = Locale(identifier: "en_US_POSIX")
    dateFormatter.timeZone = TimeZone.init(secondsFromGMT: 0)
    dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
    return dateFormatter
}()
```

Your `makeDictionary` function will look something like:

```swift
func makeDictionary(facebookId: String, accessToken: String? = nil, authenticationToken: String? = nil, expirationDate: Date) throws -> [String: String]? {

  let dateString = Self.dateFormatter.string(from: expirationDate)
  var returnDictionary = [AuthenticationKeys.id.rawValue: facebookId,
              AuthenticationKeys.expirationDate.rawValue: dateString]
  if let accessToken = accessToken {
    returnDictionary[AuthenticationKeys. accessToken.rawValue] = accessToken 
  }  else if let authenticationToken = authenticationToken {
    returnDictionary[AuthenticationKeys.token.rawValue] = authenticationToken
  } else {
    throw ParseError(code: .unknownError,
                    message: "Missing required key. Need to provide \"accessToken\" or \"authenticationToken\"")
  }
  return returnDictionary
}

```

Reply:
`CodingKeys` have been added following your suggestions. 
Additionally, the `makeDictionary` function has been modified and now `userId` and `expirationDate` are mandatory, but `accessToken` and `authenticationToken` are optional.